test/conn_test.go

```diff
@@ -2761,3 +2761,150 @@ func TestRetryOnFailedConnectWithTLSError(t *testing.T) {
 	}
 }
+
+func TestConnEventListeners(t *testing.T) {
+	t.Run("receive all events", func(t *testing.T) {
+		errs := make(chan error)
+		waitForStatusChange := func(ch chan nats.Status, expected nats.Status) {
+			t.Helper()
+			select {
+			case s := <-ch:
+				if s != expected {
+					errs <- fmt.Errorf("Invalid status received; want: %s; got: %s", expected, s)
+				}
+				errs <- nil
+			case <-time.After(10 * time.Second):
+				errs <- fmt.Errorf("Timeout waiting for event %s", expected)
+			}
+		}
+
+		s := RunDefaultServer()
+
+		var (
+			disconnected = make(chan nats.Status)
+			reconnecting = make(chan nats.Status)
+			connected    = make(chan nats.Status)
+			drainSubs    = make(chan nats.Status)
+			drainPubs    = make(chan nats.Status)
+			closed       = make(chan nats.Status)
+		)
+
+		nc, err := nats.Connect(s.ClientURL())
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		nc.RegisterStatusChangeListener(nats.DISCONNECTED, disconnected)
+		nc.RegisterStatusChangeListener(nats.RECONNECTING, reconnecting)
+		nc.RegisterStatusChangeListener(nats.CONNECTED, connected)
+		nc.RegisterStatusChangeListener(nats.DRAINING_SUBS, drainSubs)
+		nc.RegisterStatusChangeListener(nats.DRAINING_PUBS, drainPubs)
+		nc.RegisterStatusChangeListener(nats.CLOSED, closed)
+
+		go waitForStatusChange(reconnecting, nats.RECONNECTING)
+		go waitForStatusChange(connected, nats.CONNECTED)
+		go waitForStatusChange(drainSubs, nats.DRAINING_SUBS)
+		go waitForStatusChange(drainPubs, nats.DRAINING_PUBS)
+		go waitForStatusChange(closed, nats.CLOSED)
+
+		time.Sleep(100 * time.Millisecond)
+		s.Shutdown()
+		if err := <-errs; err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		if status := nc.Status(); status != nats.RECONNECTING {
+			t.Fatalf("Expected RECONNECTING status; got: %s", status)
+		}
+
+		s = RunDefaultServer()
+		defer s.Shutdown()
+		if err := <-errs; err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		if status := nc.Status(); status != nats.CONNECTED {
+			t.Fatalf("Expected CONNECTED status; got: %s", status)
+		}
+
+		if err := nc.Drain(); err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 3; i++ {
+			if err := <-errs; err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		if status := nc.Status(); status != nats.CLOSED {
+			t.Fatalf("Expected RECONNECTING status; got: %s", status)
+		}
+	})
+
+	t.Run("no not block on closed channel", func(t *testing.T) {
+		s := RunDefaultServer()
+
+		connected := make(chan nats.Status)
+
+		nc, err := nats.Connect(s.ClientURL())
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		defer nc.Close()
+		nc.RegisterStatusChangeListener(nats.CONNECTED, connected)
+		close(connected)
+
+		s.Shutdown()
+		s = RunDefaultServer()
+		defer s.Shutdown()
+		if err := nc.Publish("foo", []byte("msg")); err != nil {
+			t.Fatalf("Unexpected error")
+		}
+	})
+}
+
+func TestConnStatusChangedEvents(t *testing.T) {
```

Comment:
It was flapping because CONNECTED status was being sent twice on successful reconnects - once in `sendConnect()` and then again in `doReconnect()`. The latter seems obsolete, so I dropped sending and event at that point.